Fix apds9960_config structure to hold a const pointer to the user

MikeDK opened the pull request:
Fix apds9960_config structure to hold a const pointer to the user
allocated sensor_trigger structure.
This way the right context can be established in the callback function.

Signed-off-by: Michael Kaplan <m.kaplan@evva.com>

MaureenHelm reviewed (commented):
Thank you for the fix. Please update the commit message to address the failed compliance check:
```
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details

1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "Fix apds9960_config structure to hold a const pointer to the user"
2: B4 Second line is not empty: "allocated sensor_trigger structure."
```

MikeDK commented:
Done, but I'm not sure if the amend worked correctly ... ;)

MaureenHelm commented:
> Done, but I'm not sure if the amend worked correctly ... ;)

The amended commit message looks good, but now you need to rebase and remove the merge commits.

MikeDK commented:
O.K. -> note to self: don't press the button github is presenting me when main branch is ahead :D

I think I figured it out now, the merge commits are gone

MikeDK commented:
*bump*

needs a second reviewer ;)

gmarull reviewed (dismissed):
can you justify these changes better? Afaik, this change forces application to keep its trigger allocated.

MikeDK commented:
yes, if the application uses the trigger of the apds9960, it will have to keep its trigger object allocated ... but I think that is also true for nearly all zephyr apis which use callbacks.

The main reason for the change was that it makes no sense that the user gets the pointer of the apds9960 driver internally allocated trigger structure ... what should the user do with that pointer?

In my case I need the callback function to provide me some context - if the pointer given via the callback is the same as my allocated one, I know immediately the context - which in my case gets me the pointer of a cpp wrapper object for the apds9960 driver (because I store the trigger together with a object pointer within a struct - where I then can use CONTAINER_OF macro)

MikeDK commented:
PS: besides the context story, in the current code the trigger structure is not initialized at all ... it is just used for providing the callback function a (useless) trigger pointer

MikeDK committed:
drivers: sensors: apds9960 fix trigger callback context

In the current implementation, the apds9960 trigger callback function
is called with the pointer of a driver-internal allocated trigger
structure. This structure is not initialized anywhere, so essentially
the trigger callback gets called with junk data.

Besides the missing initialization, it would be better instead to hold
a const pointer to a user allocated sensor_trigger object.
This way user code can establish a context with related user data (for
example a pointer to a C++ object) by storing its sensor_trigger object
within a structure alongside the user data, and then using
CONTAINER_OF() macro to get the pointer of the container struct (and
thus the user data).

Signed-off-by: Michael Kaplan <m.kaplan@evva.com>

MikeDK commented:
@gmarull I changed the commit message months ago to better describe the need for the changes ... could you please re-review so this can maybe be merged? ;)

MikeDK commented:
@MaureenHelm what to do with this now? The reviewer who wanted a change does not react any more and this PR is blocked because of that :/

MaureenHelm commented:
> @MaureenHelm what to do with this now? The reviewer who wanted a change does not react any more and this PR is blocked because of that :/

Please rebase to trigger a new CI run.

MikeDK commented:
@MaureenHelm rebase done ;)

MikeDK commented:
@avisconti @teburd maybe some time for a small review? ;)

MikeDK commented:
stale again :/

PR is open half a year now, and only @MaureenHelm managed to do a review (thanks btw!) ... Would it be too harsh to say that I am a little disappointed?

We would really like to use zephyr-rtos for the next big project of our company, but if it is so hard to get PRs merged, then we might do a little rethinking :/
